engines/sci/graphics/palette32.cpp

```diff
@@ -21,20 +21,156 @@
+		// it be left the way it is already?
+		assert(kernelSetFromResource(999, true));
+	}
 
 GfxPalette32::~GfxPalette32() {
 	unloadClut();
+	varyOff();
 	cycleAllOff();
 }
 
+void GfxPalette32::submit(Palette &palette) {
+	// TODO: This code won’t do much unless/until submit can
+	// receive HunkPalettes in addition to SOLPalettes. In SCI engine,
+	// SOLPalettes do not have timestamp/version fields, only
+	// HunkPalettes do, but in ScummVM there is only one type of palette.
+	if (palette.timestamp == _version) {
+		return;
+	}
+
+	// TODO: Verify this is doing a copy.
+	Palette oldSourcePalette(_sourcePalette);
+
+	// TODO: In SCI32 this is actually calling SOLPalette::merge,
+	// so there should probably be a merge function on Sci::Palette
+	// instead of doing it inline here like this.
+	for (int i = 0; i < ARRAYSIZE(_sourcePalette.colors); ++i) {
+		if (palette.colors[i].used) {
+			_sourcePalette.colors[i] = palette.colors[i];
+		}
+	}
+
+	if (!_needsUpdate && _sourcePalette != oldSourcePalette) {
+		++_version;
+		_needsUpdate = true;
+	}
+
+	// TODO: Technically this information is supposed to be
+	// persisted through a HunkPalette object; right now it
+	// would just be lost once the temporary palette was destroyed.
+	// palette.timestamp = _version;
+}
+
+override bool GfxPalette32::kernelSetFromResource(GuiResourceId resourceId, bool force) {
+	// TODO: In SCI32, palettes that come from resources come in as
+	// HunkPalette objects, not SOLPalette objects. The HunkPalettes
+	// have some extra persistence stuff associated with them, such that
+	// when they are passed to GfxPalette32::submit, they would get the
+	// version number of GfxPalette32 assigned to them.
+	Palette palette;
+
+	if (createPaletteFromResourceInternal(resourceId, &palette)) {
+		submit(palette);
+		return true;
+	}
+
+	return false;
+}
+
+// TODO: This is wrong; it is just overridden for the time being to put stuff
+// on _nextPalette instead of _sysPalette
+override void GfxPalette32::set(Palette *newPalette, bool force, bool forceRealMerge) {
+	submit(*newPalette);
+}
+
+void GfxPalette32::updateForFrame() {
+	applyAll();
+	_needsUpdate = false;
+	// _frameOut.remap.remapAllTables(_nextPalette != _sysPalette);
+}
+
+void GfxPalette32::updateHardware() {
+	if (_sysPalette == _nextPalette) {
+		debug("Skipping hardware update because palettes are identical");
+		return;
+	}
+
+	byte bpal[3 * 256];
+
+	// TODO: Copy assignment operator for Palette??
+	for (int i = 0; i < ARRAYSIZE(_sysPalette.colors); ++i) {
+		_sysPalette.colors[i] = _nextPalette.colors[i];
+
+		// NOTE: If the brightness option in the user configuration file is set,
+		// SCI engine adjusts palette brightnesses here by mapping RGB values to values
+		// in some hard-coded brightness tables. There is no reason on modern hardware
+		// to implement this, unless it is discovered that some game uses a non-standard
+		// brightness setting by default
```

Comment:
Perhaps we could implement that as an option in scummvm.ini? Though, I personally don't think it's worth the effort, so we could just leave it as-is